pkg/cache/v3/snapshot.go

```diff
@@ -94,64 +36,32 @@ type Snapshot struct {
 }
 
 // NewSnapshot creates a snapshot from response types and a version.
-func NewSnapshot(version string,
-	endpoints []types.Resource,
-	clusters []types.Resource,
-	routes []types.Resource,
-	listeners []types.Resource,
-	runtimes []types.Resource,
-	secrets []types.Resource,
-	extensionConfigs []types.Resource) Snapshot {
-	return NewSnapshotWithResources(version, SnapshotResources{
-		Endpoints:        endpoints,
-		Clusters:         clusters,
-		Routes:           routes,
-		Listeners:        listeners,
-		Runtimes:         runtimes,
-		Secrets:          secrets,
-		ExtensionConfigs: extensionConfigs,
-	})
-}
-
-// SnapshotResources contains the resources to construct a snapshot from.
-type SnapshotResources struct {
-	Endpoints        []types.Resource
-	Clusters         []types.Resource
-	Routes           []types.Resource
-	Listeners        []types.Resource
-	Runtimes         []types.Resource
-	Secrets          []types.Resource
-	ExtensionConfigs []types.Resource
-}
-
-// NewSnapshotWithResources creates a snapshot from response types and a version.
-func NewSnapshotWithResources(version string, resources SnapshotResources) Snapshot {
+// The resources map is keyed off the type URL of a resource, followed by the slice of resource objects.
+func NewSnapshot(version string, resources map[resource.Type][]types.Resource) Snapshot {
 	out := Snapshot{}
-	out.Resources[types.Endpoint] = NewResources(version, resources.Endpoints)
-	out.Resources[types.Cluster] = NewResources(version, resources.Clusters)
-	out.Resources[types.Route] = NewResources(version, resources.Routes)
-	out.Resources[types.Listener] = NewResources(version, resources.Listeners)
-	out.Resources[types.Runtime] = NewResources(version, resources.Runtimes)
-	out.Resources[types.Secret] = NewResources(version, resources.Secrets)
-	out.Resources[types.ExtensionConfig] = NewResources(version, resources.ExtensionConfigs)
+
+	for typ, resource := range resources {
+		index := GetResponseType(typ)
+		if index != types.UnknownType {
+			out.Resources[index] = NewResources(version, resource)
+		}
+	}
 
 	return out
 }
 
-func NewSnapshotWithTtls(version string,
-	endpoints []types.ResourceWithTTL,
-	clusters []types.ResourceWithTTL,
-	routes []types.ResourceWithTTL,
-	listeners []types.ResourceWithTTL,
-	runtimes []types.ResourceWithTTL,
-	secrets []types.ResourceWithTTL) Snapshot {
+// NewSnapshotWithTTLs creates a snapshot with time to live per resource.
```

Comment:
I would just phrase this as `creates a snapshot of ResourceWithTTLs` or something, people can refer to ResourceWithTTL to understand what fields can be set

Reply:
Cool that sounds good